GPS subclasses, requirements + code cleansing

Ptosiek opened the pull request:
Allow setting port and host on GPSD for test purpose.
It keeps i2c/uart for now

I know there are conflicts because of the recent changes you made to the gadgetbride implementation. I need to have a look at it, and will prob have questions (maybe you can explain why these changes ?)

I'm opening it already so we can start discussion it anyway. Especially the part where the time is set, where we are running "uname". I don't really get why it's done that way. (Also the utility of timestamp_done in GB)

Ptosiek commented:
For the GB changes, the questions are:
1. Why the need of: decode("utf-8", "ignore")?
2. Why the change to re.match(...)**[1:]**?
3. Why do you need access to timestamp_bytes in this log ? app_logger.info(f"{len(self.value)}, {len(self.timestamp_bytes)}")

(this does not exist in the code anymore since I've simplified the handling of frame)

hishizuka commented:
Hi @Ptosiek ,
sorry to let you fix the conflict, I didn't realize your PR included a gadgetbridge fix.

Please give me a little time to check as the functions around GB are not working. It was killed by around `get_utc_time_from_datetime` and also by `show_message`(message["t"] == "notify").

When an error occurs in `rx_characteristic`, the program is killed with a large error log. This has been happening since the introduction of the logger, and the old behavior was that a single eventloop would stop and the program would not be killed.

Here is a quick reply.

1. This is to reduce errors in rx_characteristic mentioned above. In my environment, UTF multi-byte (Japanese) decoding sometimes fails.
2. This is because 0x10 is considered the first character. However, `message.startswith` seems to have the behavior of ignoring the 0x10 at the beginning.
3. It appears that the debug code was pasted by mistake and committed as is. Not necessary.

Ptosiek commented:
Sorry the new code was not supposed to work yet (tough it kinda was before, I had broken it :) and was waiting for your answers.

I've rebased the branch master with your changes and fixed it, it should be working now (though I tested only sending gps data, not sure how to have notification).
Not that I've dropped the previous update_manual in favor of having the gadgetBridge related code in one place only.

=> Also, when an error occurs in rx_characteristic, the program is killed with a large error log
Yes I've seen that. I've added a catch all error in decode_message that seems to prevent that from happening?

My questions around setting the time (cf TODO in the code) are still actual though

hishizuka commented:
Hi, @Ptosiek ,
in GadgetbridgeService, the config assignment are missing.
Cannot call self.config.gui.show_message.

For judging the first string of L80, F_BYTE_MARKER should take into account that 0x10 happens to be at the beginning of the data in the middle divided by 20 bytes (it is possible with multi-byte such as Japanese), and add `self.message is not None` or simply `not self.message`.
The same applies to the determination of the trailing string. It is more robust to set L_BYTE_MARKER to `)¥n` or [`))¥n` for setTime or `})¥n` for GB] since `\n` can be inserted in the middle of the message. It is possible that `\n` could happen to come at the end of the 20 bytes in the middle of the message.

I will look at the other files.

hishizuka commented:
>  not sure how to have notification
After turning on all permissions during installation, connect within the GB, return to the home screen, and press the power button to sleep.
In that state, if you receive GMail or any message application, the notification should be forwarded.

Ptosiek commented:
1. in GadgetbridgeService, the config assignment are missing.
Cannot call self.config.gui.show_message
Have to think about it. Trying to make the object independent here

2.  For end/start marker, I am not sure. None of the solutions seems very robust so need to look deeper
Can you share an example of a frame (in Japanese) you have ?

Ptosiek commented:
So, I've fixed the passing the gui to display message and handling the failure of 'atob' decoding.
Though you maybe were still displaying the notification in that case, but with empty content?

About the markers, I don't think the cases you mention can happen.
The data is being escaped and encoded first. cf. this code:
https://codeberg.org/Freeyourgadget/Gadgetbridge/src/branch/master/app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/banglejs/BangleJSDeviceSupport.java
I tried to send several messages (through the send/debug function) with random Japanese and Russian words and It always end up being 'atob' encoded (with 'safe' characters) and well.. it's actually unusable since it seems GB will use ISO_8859_1 anyway.
Dud you manage to display content in non latin script ?

hishizuka commented:
Hi @Ptosiek ,
yes, they had recently put in atob. Thanks very much.

Since GadgetbridgeService is very useful to do,

1. I would like to implement track in the future where it handles GPS data (`m_type == "gps":`). Actually, get_index does not work without track. Omitting the GPS module is good for beginners.
2. In the future, I would like to make the receiving part of GadgetbridgeService an independent library. I would also like to integrate the Apple Notification Center Service (ANCS) on the Apple side.

https://learn.adafruit.com/ancs-gizmo/overview
https://twitter.com/search?q=adafruit%20clue%20bike&src=typed_query&f=top

hishizuka commented on reqs/min.in:
> qsync==0.24.0

You are watching carefully, thank you!

hishizuka commented on modules/sensor/gps/base.py:
I agree with dropping dateutil.

for tz,

L4

- (-) from datetime import datetime, time, timedelta
- (+) from datetime import datetime, time, timezone

L313
- (-) kernel_date = datetime(2019, 1, 1, 0, 0, 0, 0, tz.tzutc())
- (+) kernel_date = datetime(2019, 1, 1, 0, 0, 0, 0, timezone.utc)

for isoformat,
as you pointed out, it is the caller of this function that is responsible for creating isoformat.

hishizuka replied:
In a desktop environment without GPS, the import of L8 timezonefinder causes an import error, so I would like it to be moved into the try statement as before.
If set_timezone can't import either, it's immediately return False.

Ptosiek replied:
In a desktop environment without GPS, the import of L8 timezonefinder causes an import error, so I would like it to be moved into the try statement as before.
If set_timezone can't import either, it's immediately return False.
=> I had put the timezonefinder in min.in requirements which represents the minimum requirements for the app to run.
Since there's not really 'no GPS' option (since even with gadgetBridge it becomes necessary).

About the time though, my question was about why are we getting the kernel_date time and comparing and why are we doing this with uname instead of calling date ?

hishizuka replied:
Indeed, since GB was added this year, the handling of timezonefinder needs to be changed. Thanks.

The reason kernel_date is needed is to check the time immediately after GPS fixes. There is a module that returns dates in the 2000s.

I found the regular expression was wrong. I will fix it.
case1: "#1253 Thu Aug 15 11:37:30 BST 2019" ... OK
case2: "#1642 SMP Mon Apr 3 17:20:52 BST 2023" ... NG (day was not 0 filled)

hishizuka commented on modules/logger_core.py:
L490

> if v in [self.config.G_GPS_NULLVALUE, self.config.G_ANT_NULLVALUE]:

This cannot be changed yet.
Since all G_GPS_NULLVALUE are not yet set to np.nan, the "n/a" will come at the time of GPSd.

Ptosiek replied:
Well this should change the n/a to None
```
        # our first task is to align the format for each null value, we do it only for GPS
        # that do not use None as NULL_VALUE already, else the values are already correct.
        # (Maybe it could be done on the sensor implementation itself)
        if self.NULL_VALUE is not None:

            def id_or_none(value):
                if isinstance(value, list):
                    return [id_or_none(x) for x in value]
                else:
                    return value if not self.is_null_value(value) else None

            lat = id_or_none(lat)
            lon = id_or_none(lon)
            alt = id_or_none(alt)
            speed = id_or_none(speed)
            track = id_or_none(track)
            mode = id_or_none(mode)
            error = id_or_none(error)
            dop = id_or_none(dop)
            # no need to check for satellites, manually computed
 ```
 But it was actually a point to discuss indeed because we mix None and np.nan here. And wanted to check it.
 
 (Btw There's no hurry to merge the PR, we can discuss and solve all the points before)

hishizuka replied:
OK, 
so you were converting to None here.
My idea is that the missing values should eventually be unified between sensor(ANT+, I2C, GPS) to np.nan. As I recall, I2C and ANT+ are already np.nan. the GPS library has a different behavior of outputting missing values, which is why the code is the way it is now.

From now on, I will let you PEND with the review function.
(Sorry I've never used it before and am unfamiliar with it)

Thanks!

hishizuka commented:
The bluetooth area in linux is not well documented and quite difficult. 

I am moving to dbus_next which can use asyncio instead of dbus, but I am not good enough and am still relying on bluez_peripheral to create a peripheral. This is even more difficult as none of the documentation is available. The only legitimate way is to look at both the [bluez documentation](https://github.com/bluez/bluez/tree/master/doc) and the [dbus_next documentation](https://python-dbus-next.readthedocs.io) to understand it. But it is difficult.
